contrib/gorm.io/gorm/gorm.go

```diff
@@ -0,0 +1,168 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-2020 Datadog, Inc.
+
+// Package gorm provides helper functions for tracing the gorm.io/gorm package (https://github.com/go-gorm/gorm).
+package gorm
+
+import (
+	"context"
+	"math"
+	"time"
+
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/ext"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
+
+	"gorm.io/gorm"
+)
+
+type key string
+
+const (
+	gormConfigKey        = key("dd-trace-go:config")
+	gormSpanStartTimeKey = key("dd-trace-go:span")
+)
+
+// Open opens a new (traced) database connection. The used driver must be formerly registered
+// using (gopkg.in/DataDog/dd-trace-go.v1/contrib/database/sql).Register.
+func Open(dialector gorm.Dialector, cfg *gorm.Config, opts ...Option) (*gorm.DB, error) {
+	db, err := gorm.Open(dialector, cfg)
+	if err != nil {
+		return db, err
+	}
+
+	return WithCallbacks(db, opts...)
+}
+
+// WithCallbacks registers callbacks to the gorm.DB for tracing.
+// It should be called once, after opening the db.
+// The callbacks are triggered by Create, Update, Delete,
+// Query and RowQuery operations.
+func WithCallbacks(db *gorm.DB, opts ...Option) (*gorm.DB, error) {
```

Comment:
The documentation for `WithCallbacks` implies that it should be called after `Open`, but `Open` already calls `WithCallbacks`.

We may not even want to export `WithCallbacks`. It is only useful in a case when the user can't change the `gorm.Open` call, in which case they're also probably not able to call `sqltrace.Open` to pass into the dialector config. What do you think? I'm leaning towards making it private.

If we are going to keep `WithCallbacks`, we should:
1. Make the documentation clear that it should only be called on a `*gorm.DB` opened with `gorm.Open`, not `gormtrace.Open`.
2. Add an example for using `WithCallbacks`, since this is not supposed to be used with `gormtrace.Open`.